src/modules/ekf2/EKF/aid_sources/magnetometer/mag_control.cpp

```diff
@@ -185,74 +187,47 @@ void Ekf::controlMagFusion()
 
 		// if we are using 3-axis magnetometer fusion, but without external NE aiding,
 		// then the declination must be fused as an observation to prevent long term heading drift
-		// fusing declination when gps aiding is available is optional.
-		const bool not_using_ne_aiding = !_control_status.flags.gps && !_control_status.flags.aux_gpos;
-		_control_status.flags.mag_dec = _control_status.flags.mag
-						&& (not_using_ne_aiding || !_control_status.flags.mag_aligned_in_flight);
+		const bool no_ne_aiding_or_pre_takeoff = !using_ne_aiding || !_control_status.flags.in_air;
+		_control_status.flags.mag_dec = _control_status.flags.mag && no_ne_aiding_or_pre_takeoff;
 
 		if (_control_status.flags.mag) {
 
 			if (continuing_conditions_passing && _control_status.flags.yaw_align) {
 
-				if (mag_sample.reset || checkHaglYawResetReq() || wmm_updated) {
+				if (mag_sample.reset || checkHaglYawResetReq() || (wmm_updated && no_ne_aiding_or_pre_takeoff)) {
```

Comment:
TODO: re-evaluate checkHaglYawResetReq() on real vehicles

At a minimum skip it if unnecessary, but it should also be possible to actually enter full mag_3D preflight.

Reply:
The problem is that it's hard to tell if the reset is required before it's too late. When using GNSS aiding, resetting the yaw variance helps a lot to prevent toilet bowling if the heading error is < 90 degrees

Reply:
What I don't like is that you can't get into full mag_3d fusion until in flight and the somewhat arbitrary hard coded 1.5m reset. You could be flying around for a while, rotating, etc and then finally poke up above 1.5 and trigger the reset.

Maybe the reset logic could compute the new mag_I/mag_B/heading and only commit (and reset covariances) if there's a sufficient difference.